Merge spring2025

Maillman opened the pull request:
Merge spring2025 into main.

19mdavenport commented:
@frozenfrank Did I forget anything?

frozenfrank commented:
Resolves #29.

frozenfrank commented:
> @frozenfrank Did I forget anything?

@19mdavenport No. Everything that I'm aware of has been covered. Thanks for doing such great work!

leesjensen reviewed (approved):
Looks great. Thanks!

19mdavenport committed:
fix nullability issues

frozenfrank committed:
Fix whitespace in test cases...

This completes a change begun in commit 55b0f1a.

frozenfrank committed:
Rename and adjust test names...

- Clarify expected outcome
- Align test with actual behavior
- Adjust behavior of 'Cannot Castle Through Check' to _only_ test this behavior

frozenfrank committed:
Remove duplication in king position and castle assertions

frozenfrank committed:
Refactor board creation helper...

This also fixes a few cases where we were not
explicitly calling 'setTeamTurn()' before operating on the board.

This This change is consistent with 90636f494f29c09afa5079649b91eaaf6123291f.

frozenfrank committed:
Add uncovered tests for FROM check and INTO check

frozenfrank committed:
Standardize test function names

frozenfrank committed:
Extract makeMoveAndAssertBoard()

frozenfrank committed:
Switch the order of king/queen side parameters for assertBlackCanCastle...

This change reflects the fact that we always draw the board from WHITE's
perspective in the boardText and other illustrations in the test.

frozenfrank committed:
Reorder and explicitly order the tests to align with written requirements...

The numerical value assigned to each order matches the it's number
in the spec document:

0. Castling works
1. No castling after moves
2. No castling through pieces
3. No castling from/through/in check

frozenfrank committed:
Rename chess/piece ==> chess/piecemoves tests... Resolves #15

frozenfrank committed:
Create ChessPieceTests with copied equals methods

frozenfrank committed:
Rename combined equals & hashcode testing for clarity

frozenfrank committed:
Introduce EqualsTestingUtility and install in ChessPieceTests

frozenfrank committed:
Simplify and improve EqualsTestingUtility interface

frozenfrank committed:
Add thoroughness to ChessPieceTests allDifferent

frozenfrank committed:
Improve debugability of EqualsTestingUtility...

- Add class comment describing role of class
- Require className to be provided as context
- Use className to make error messages more specific
- Add error messages to all remaining assertions

frozenfrank committed:
Install EqualsTestingUtility in ChessMoveTests

frozenfrank committed:
Install EqualsTestingUtility in ChessPositionTests

frozenfrank committed:
Install EqualsTestingUtility in ChessBoardTests

frozenfrank committed:
Move 'Piece Move on All Pieces' to ChessPieceTests...

In order to break the dependency on ChessBoard.resetBoard(),
this also switches the test to simply evaluating pieceMoves()
on all possible piece types on all possible squares on the board.

768 total evaluations is a high number, but it also gives us
strong confidence that the code never throws errors!

frozenfrank committed:
Provide meaningful different boards for hashCode testing

frozenfrank committed:
Add 'Construct Empty ChessBoard' test

frozenfrank committed:
Fix issue with failing ChessBoard tests...

The root cause stemmed from the probability that a piece
would be replaced with the exact same piece. When this happens,
the board is not actually different, and test would fail improerly.

Since there are 6 pieces * 2 colors = 12 pieces, the probabiliy
of an individual piece being replaced with itself is 1 in 12.
This is about 9%. However, there are 14 different pieces that
each have this probability. The total probability compounds
up to ~66%!

Because of this overlooked error, the tests would fail improperly
most of the time.

This change does not call resetBoard() before creating each of the boards.
Therefore, they are all blank, and adding a random piece in a different
position each time always creates a definitely different board.

frozenfrank committed:
Reorder ChessBoard tests to read in an order students are more likely to implement them

frozenfrank committed:
Avoid creating impossible chess scenarios...

- Prevent placing pawns on the end rows
- Remove pawn promotions to kings & pawns

While I personally think that lower-level classes should not be making
these kinds of assumptions, I also accept that students sometimes do.

Consistent with the change in 38c4f14487b483ee9ac922cc124babc63576fe03,
we want to allow students to make more assumptions than they should.

frozenfrank committed:
Remove randomness and pawns from back row

frozenfrank committed:
Merge branch 'chess-position-tests' into chess-board-test-improvements

frozenfrank committed:
Fix array index issue and improve readability

frozenfrank committed:
Add ChessGameTests...

These apply the EqualsTestingUtility class to ChessGame.
Although the equals() and hashCode() methods aren't
_strictly_ necessary for future phases, this is the
easiest, most reasonable and consistent way to ensure
that no fields are defined as 'static'.

Additionally, it doesn't hurt to have students
remember how to generate the two methods again.

Adding another test here isn't anymore of a burden
on students getting started on Phase 1 because we
already instruct them to copy over the entire
'game/' directory in the instructions.

jerodw committed:
Merge pull request #13 from softwareconstruction240/1-update-castling-tests-with-clarified-definition-of-valid-check

Update castling tests with clarified definition of valid check

jerodw committed:
Merge pull request #16 from softwareconstruction240/chess-position-tests

Chess Piece tests

jerodw committed:
Merge pull request #17 from softwareconstruction240/chess-board-test-improvements

Chess board test improvements

frozenfrank committed:
Merge remote-tracking branch 'origin/next-spring2025' into null-fix

jerodw committed:
Merge pull request #10 from softwareconstruction240/null-fix

Fix Phase 0 tests: nullability issues

19mdavenport committed:
fix isInCheck error messages

frozenfrank committed:
Use defined variables for error messages...

Although this was the only discovered omission,
this commit represents an exhaustive search through
all the testing files to verify comprehensiveness.

All places which should be using these variables
are properly referencing them.

jerodw committed:
Merge pull request #21 from softwareconstruction240/fix-phase1-test-error-messages

Fix Phase 1 Tests IsInCheck Error Messages

frozenfrank committed:
Add LoginBadRequest test

frozenfrank committed:
Rename tests for consistency

frozenfrank committed:
Divide test file into large sections

frozenfrank committed:
Condense lines to stay within 500 lines

frozenfrank committed:
Combine similar loginUnauthorized tests

jerodw committed:
Merge pull request #19 from softwareconstruction240/chess-game-tests

Add ChessGameTests

jerodw committed:
Merge pull request #22 from softwareconstruction240/standardize-phase3-tests

Standardize phase3 tests

mewilker committed:
edit DatabaseManager.java to remove DBInfo

mewilker committed:
remove excessive external try from code that loads db.properties

mewilker committed:
edit DatabaseManager.createDatabase to use a try with resources on it's connection

mewilker committed:
edit DatabaseManager.createDatabase to be public so students by default can use it in their Server class

mewilker committed:
added a code comment in DatabaseManager.getConnection warning students not to wrap the Driver's getConnection with a try with resources

mewilker committed:
remove getConnection parameter that is not required

mewilker committed:
rephrase the comment so it's more clear that this is the only place to not use a try-with-resources

mewilker committed:
fix spacing DatabaseManager.createDatabase to match Intellij suggestions

frozenfrank committed:
Improve exception throwing in DatabaseManager

This requires replacing the constructor in DataAccessException.

This has the unintended side-effect of also making it harder for
students to unintentionally use this exception for their own
custom exceptions where they should really be creating their
new custom exception classes.

frozenfrank committed:
Standardize error throwing...

All changes are in DatabaseManager, but all other starter-code files
were checked for consistency in the process. DatabaseManager is the
only starter code file that performs any explicit error catching/throwing.

Standardizations:
- Always name the exception variable 'ex'
- Use lowercase in error message
- Always include the exception in the wrapped throwable
  This also applies to the RuntimeException
- Adjust error handling string management to better handle
  the entire field of potential errors that could be thrown

frozenfrank committed:
Update DataAccessException.java Restore String only constructor

This fulfills the request by @jerodw.

Currently, this new constructor is unused. 
@19mdavenport Will this cause any issues with the 
style checker and unused code?

mewilker committed:
specifically state the db.properties file not found in error message

frozenfrank committed:
Revert "specifically state the db.properties file not found in error message"

This reverts commit c6282dfec2e5a63e4b21fb2334ef3bdedc6b5589.

The intended benefit is already covered by the 'catch' clause
on line 30 which rethrows the exception (and other exceptions too)
with the specific name of the file.

19mdavenport committed:
adjust WebSocketTests improve error messages

jerodw committed:
Merge pull request #23 from softwareconstruction240/phase-4-fixes

Edit DatabaseManager.java to improve try blocks and documentation

19mdavenport committed:
adjust WebSocketTests add expected types to incorrect size message (reccomendation from a student)

TheDavSmasher committed:
Added Create Bag Request test

I was going to add a couple more tests but realized that James made basically the same tests I already had with different names and other minor changes. This was the only test I had unique/new so I'm making a PR for it

jerodw committed:
Merge pull request #26 from softwareconstruction240/standard-api-tests-include-bad-create

Added Create Bad Request test

jerodw committed:
Merge pull request #25 from softwareconstruction240/websocket-test-error-messages

Improve WebSocket Tests Error Messages

19mdavenport committed:
add new test(DatabaseTests) server gives 500 when database cannot be contacted

19mdavenport committed:
fix nested try/catch

jerodw committed:
Merge pull request #28 from softwareconstruction240/database-500-test

DatabaseTests Server Error/500 Test

mewilker committed:
If TestFactory.java detects a debug environment, it will increase test time

jerodw committed:
Merge pull request #29 from softwareconstruction240/test-factory-debug-mode

Test Factory detects Debug mode
